Add 'diagnose datadog-connectivity' subcommand

Kaderinho opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?

This PR transforms the `diagnose` simple command into a command that has 2 subcommands which are 

- `datadog-connectivity`
- `metadata-availability`

`metadata-availability` is doing the same job as the `diagnose` command.

`datadog-connectivity` is a new hidden command which aims to troubleshoot connectivity issues between the Agent and Datadog, especially when using a proxy.

It is just an empty function for now but it will be filled up in future PRs.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succint form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

Troubleshooting connectivity issues is hard so we want to add some tools in the Agent to help us on that.

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

Decided to create a subcommand because I believed that the `diagnose` name correctly represents the purpose of this new subcommand and also to avoid creating too many commands.

The choice of the subcommand names is not set in stone so feel free to suggest others names if you think they are more relevant.

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

Disclaimer : This is the first PR of a long list of PRs that aim to add connectivity troubleshooting tools so it's might not be up to date during the QA. Don't hesitate to ping me during the QA if there is any problem.

Simply verify that : 

- `datadog-agent diagnose` works like before.
- `datadog-agent diagnose metadata-availability` works and does the same things as `datadog-agent diagnose`.
- `datadog-agent diagnose datadog-connectivity` exists and outputs something.


<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

Kaderinho committed:
feat(diagnose): add empty new diagnose subcommand 'datadog-connectivity'

Signed-off-by: Nicolas Guerguadj <nicolas.guerguadj@epita.fr>

Kaderinho committed:
feat(diagnose): move original 'diagnose' command to a subcommand

With the new 'datadog-connectivity' subcommand for 'diagnose' command, we want
to be able to detect typos when launching the command.

For instance if we call 'diagnose blabla', the 'diagnose' command will run,
because 'blabla' is considered as a flag, while we want to display
an error / help message.

By moving the metadata availability checks to a subcommand we are able to
correctly display the help information.

Now instead of using 'datadog-agent diagnoe' we have to use
'datadog-agent diagnose metadata-availability'

Signed-off-by: Nicolas Guerguadj <nicolas.guerguadj@epita.fr>

Kaderinho committed:
chore(releasenotes): add a release note for the diagnose command and new subcommands

Signed-off-by: Nicolas Guerguadj <nicolas.guerguadj@epita.fr>

alai97 commented on releasenotes/notes/create-diagnose-datadog-connectivity-subcommand-53403e7599a38084.yaml:
```diff
@@ -0,0 +1,17 @@
+---
+upgrade:
+  - |
+    ``diagnose`` command has now subcommands. The original ``diagnose``
+    behavior was transfered to ``diagnose metadata-availability`` subcommand.
```
```suggestion
    The ``diagnose`` command has subcommands. The original ``diagnose``
    behavior was transferred to a ``diagnose metadata-availability`` subcommand.
```

Kaderinho replied:
I removed this part as there is no breaking change anymore.

alai97 commented on releasenotes/notes/create-diagnose-datadog-connectivity-subcommand-53403e7599a38084.yaml:
```diff
@@ -0,0 +1,17 @@
+    behavior was transfered to ``diagnose metadata-availability`` subcommand.
+features:
+  - |
+    The new ``diagnose datadog-connectivity`` subcommand allows you to
+    get information about the state of the connectivity between the Agent
+    and Datadog endpoints.
```
```suggestion
    The ``diagnose datadog-connectivity`` subcommand allows you to
    get information about the state of the connectivity between the Agent
    and Datadog endpoints.
```

Kaderinho replied:
We will remove release note for now as the feature is not done yet.

alai97 reviewed (approved):
Looks good, two release note nits!

Kaderinho committed:
feat(diagnose): add backward compatibility on diagnose command

Signed-off-by: Nicolas Guerguadj <nicolas.guerguadj@epita.fr>

Kaderinho committed:
fix(diagnose): make 'diagnose datadog-connectivity' a hidden command

This command is doing nothing for now so let's hide it until it does
useful things

Signed-off-by: Nicolas Guerguadj <nicolas.guerguadj@epita.fr>

Kaderinho committed:
docs(diagnose): remove update part in the release note as there is no breaking change anymore

Signed-off-by: Nicolas Guerguadj <nicolas.guerguadj@epita.fr>

Kaderinho commented on cmd/agent/app/diagnose.go:
```diff
@@ -16,18 +16,37 @@ import (
+		Short: "Execute some connectivity diagnosis on your system",
+		Long:  ``,
+	}
+
```
Done in    [f63b3581acb](https://github.com/DataDog/datadog-agent/pull/11913/commits/f63b3581acb33aebfcb0faf6d821ba2c4db1f5a0 )

clamoriniere replied:
another nits:
you might be able to send to `stderr` a warning saying this command is deprecated.
I think the cobra framework allows it

Kaderinho replied:
Done [here](https://github.com/DataDog/datadog-agent/pull/11913/commits/284c69ce3ab1efaf43fb070732553816dafd935b)

Kaderinho replied:
We are not deprecating it for now so I removed the message

djmitche reviewed (approved):
I like that this keeps the existing default behavior of `agent diagnose` while allowing more specific diagnoses.  It'd be cool if we could do the same with `agent status` (e.g., `agent status logs`)!

Kaderinho committed:
chore(diagnose): add deprecation message for diagnose command

Note that Cobra is adding "Command 'diagnose' is deprecated, " before the string in the code

Signed-off-by: Nicolas Guerguadj <nicolas.guerguadj@epita.fr>

vickenty commented on cmd/agent/app/diagnose.go:
```diff
@@ -16,18 +16,40 @@ import (
+
+	diagnoseMetadataAvailabilityCommand = &cobra.Command{
+		Use:   "metadata-availability",
+		Short: "Execute some availability checks on cloud provider metadata endpoints",
```
```suggestion
		Short: "Check availability of cloud provider metadata endpoints",
```

vickenty commented on cmd/agent/app/diagnose.go:
```diff
@@ -16,18 +16,40 @@ import (
+var (
+	diagnoseCommand = &cobra.Command{
+		Use:        "diagnose",
+		Short:      "Execute some connectivity diagnosis on your system",
```
```suggestion
		Short:      "Check availability of cloud provider metadata endpoints",
```

vickenty commented on cmd/agent/app/diagnose.go:
```diff
@@ -16,18 +16,40 @@ import (
+		Use:        "diagnose",
+		Short:      "Execute some connectivity diagnosis on your system",
+		Long:       ``,
+		Deprecated: "this command is only here for backward compatibility. Please consider using 'diagnose metadata-availability' instead.",
```
Deprecation implies backward compatibility already. Also would be good to mention deprecation policy: is it going to be removed in Agent 7.38? Agent 8.0? Never (in which case we might want to drop deprecation, and simply support it as an alias).

```suggestion
		Deprecated: "Renamed to 'diagnose metadata-availability'.",
```

Kaderinho replied:
I don't think we will remove this command in the future. Maybe we'll even add behavior to this command to execute both subcommand successively idk 🤷

vickenty commented on cmd/agent/app/diagnose.go:
```diff
@@ -16,18 +16,40 @@ import (
+
+	diagnoseDatadogConnectivityCommand = &cobra.Command{
+		Use:    "datadog-connectivity",
+		Short:  "Execute some connectivity diagnosis between your system and Datadog endpoints",
```
```suggestion
		Short:  "Check connectivity between your system and Datadog endpoints",
```

vickenty commented on releasenotes/notes/create-diagnose-datadog-connectivity-subcommand-53403e7599a38084.yaml:
```diff
@@ -0,0 +1,13 @@
+#
+# Each section note must be formatted as reStructuredText.
+---
+features:
```
If the feature is not implemented yet, we probably don't want to announce it in the changelog. We should either remove this (and use `no-changelog` label), or implement the command before merging this PR.

Kaderinho committed:
Update cmd/agent/app/diagnose.go

Co-authored-by: Vickenty Fesunov <vickenty@users.noreply.github.com>

Kaderinho committed:
Update cmd/agent/app/diagnose.go

Co-authored-by: Vickenty Fesunov <vickenty@users.noreply.github.com>

Kaderinho committed:
Update cmd/agent/app/diagnose.go

Co-authored-by: Vickenty Fesunov <vickenty@users.noreply.github.com>

Kaderinho committed:
fix(releasenote): remove release note as the feature is not ready yet

Signed-off-by: Nicolas Guerguadj <nicolas.guerguadj@epita.fr>

Kaderinho commented:
I don't think we need to deprecate the `diagnose` command for now. Maybe we can use it to run all diagnostics when `datadog-connectivity` will be implemented. What do you think ? @vickenty @clamoriniere @olivielpeau

olivielpeau commented:
Agreed that we shouldn't deprecate it for now.

We can decide whether to deprecate it later. Just having it run all the diagnostics instead of deprecating it would make sense to me as well.

vickenty reviewed (approved):
LGTM

Kaderinho committed:
fix(diagnose): remove deprecation message for diagnose command

Signed-off-by: Nicolas Guerguadj <nicolas.guerguadj@epita.fr>

olivielpeau commented on cmd/agent/app/diagnose.go:
```diff
@@ -16,18 +16,39 @@ import (
+
+	diagnoseMetadataAvailabilityCommand = &cobra.Command{
+		Use:   "metadata-availability",
+		Short: "Check availability of cloud provider metadata endpoints",
```
@Kaderinho I think we should rename that description a little bit, since the command diagnoses not only cloud provider metadata, but also other sources, such as containerd, docker, kubelet, etc. So let's rename the description to something like `Check availability of cloud provider and container metadata endpoints`?

(same comment on line 22, and this isn't urgent and can be addressed for 7.38)

Kaderinho replied:
Created the PR here https://github.com/DataDog/datadog-agent/pull/12118
